src/kernel/MomentumOpenAdvDiffElemKernel.C

```diff
@@ -0,0 +1,370 @@
+      w_uspecBip[j] = 0.0;
+      w_coordBip[j] = 0.0;
+      const double axj = vf_exposedAreaVec(ip,j);
+      asq += axj*axj;
+    }
+    const DoubleType amag = std::sqrt(asq);
+    
+    // interpolate to bip
+    DoubleType viscBip = 0.0;
+    for ( int ic = 0; ic < NcAlgTraits::nodesPerFace_; ++ic ) {
+      const DoubleType r = vf_shape_function_(ip,ic);
+      viscBip += r*vf_viscosity(ic);
+      for ( int j = 0; j < BcAlgTraits::nDim_; ++j ) {
+        w_uspecBip[j] += r*vf_bcVelocity(ic,j);
+        w_uBip[j] += r*vf_velocityNp1(ic,j);
+        w_coordBip[j] += r*vf_coordinates(ic,j);
+      }
+    }
+    
+    // data at interior opposing face
+    for ( int ic = 0; ic < BcAlgtraits::nodesPerElement_; ++ic ) {
+      const DoubleType r = v_shape_function(ip,ic);
+      for ( int j = 0; j < BcAlgTraits::nDim_; ++j ) {
+        p_uScs[j] += r*v_velocityNp1(ic,j);
+      }
+    }
+    
+    // fixme: is this okay? Peclet factor; along the edge is fine
+    const DoubleType densL   = *stk::mesh::field_data(densityNp1, nodeL);
+    const DoubleType densR   = *stk::mesh::field_data(densityNp1, nodeR);
+    const DoubleType viscL   = *stk::mesh::field_data(*viscosity_, nodeL);
+    const DoubleType viscR   = *stk::mesh::field_data(*viscosity_, nodeR);
+    const DoubleType *uNp1R  =  stk::mesh::field_data(velocityNp1, nodeR);
+    const DoubleType *vrtmL  =  stk::mesh::field_data(*velocityRTM_, nodeL);
+    const DoubleType *vrtmR  =  stk::mesh::field_data(*velocityRTM_, nodeR);
+    
+    const DoubleType *coordL =  stk::mesh::field_data(*coordinates_, nodeL);
+    const DoubleType *coordR =  stk::mesh::field_data(*coordinates_, nodeR);
+    
+    DoubleType udotx = 0.0;
+    for ( int i = 0; i < BcAlgTraits::nDim_; ++i ) {
+      const DoubleType dxi = coordR[i]  - coordL[i];
+      udotx += 0.5*dxi*(vrtmL[i] + vrtmR[i]);
+      w_nx[i] = vf_exposedAreaVec(ip,i)/amag;
+      // extrapolation
+      DoubleType duR = 0.0;
+      for ( int j = 0; j < BcAlgTraits::nDim_; ++j ) {
+        DoubleType dxj = w_coordBip[j] - coordR[j];
+        duR += dxj*vf_Gjui(localFaceNode,i,j)*hoUpwind;
+      }
+      w_uBipExtrap[i] = uNp1R[i] + duR;
+    }
+    
+    const DoubleType diffIp = 0.5*(viscL/densL + viscR/densR);
+    const DoubleType pecFuncArg = stk::math::abs(udotx)/(diffIp+small_);
+    // FIXME: modify pecletFunction to be double type
+    DoubleType pecfac = 0.0;
+    for(int simdIndex=0; simdIndex<stk::simd::ndoubles; ++simdIndex) {
+      stk::simd::set_data(pecfac, simdIndex, pecletFunction_->execute(stk::simd::get_data(pecFuncArg, simdIndex)));
+    }
+    const DoubleType om_pecfac = 1.0-pecfac;
+   
+    //================================
+    // advection first
+    //================================
+    const DoubleType tmdot = vf_openMassFlowRate(ip);
+    
+    // advection; leaving the domain
+    // fixme... oh my, do I need to wrap this in a stk::math::if_then_else(tmdot > 0, doThis, doThat);
+    if ( tmdot > 0.0 ) {
```

Comment:
Yes you either need to use stk::math::if_then_else, or put in a loop over simd::ndoubles as is currently done above for the unconverted pecletFunction.

Reply:
I may pick the later as indents can get out of control... I think.. I will try both and see.

Reply:
Is there any cost to using this inline function for such a large code fragment? I went with the if_then_else....